crates/rnote-ui/src/dialogs/import.rs

```diff
@@ -266,21 +268,69 @@ pub(crate) async fn dialog_import_pdf_w_prefs(
-    match response {
-        ResponseType::Apply => {
+    // Listen to responses
+
+    let (tx, mut rx) = futures::channel::mpsc::unbounded::<anyhow::Result<bool>>();
+    let tx_cancel = tx.clone();
+    let tx_confirm = tx.clone();
+
+    import_pdf_button_cancel.connect_clicked(clone!(@weak dialog => move |_| {
+        dialog.close();
+
+        if let Err(e) = tx_cancel.unbounded_send(Ok(false)) {
+            tracing::error!(
+                "PDF import dialog closed, but failed to send signal through channel. Err: {e:?}"
+            );
+        }
+    }));
+
+    import_pdf_button_confirm.connect_clicked(clone!(@weak pdf_page_start_row, @weak pdf_page_end_row, @weak input_file, @weak dialog, @weak canvas => move |_| {
+        dialog.close();
+
+        let inner_tx_confirm = tx_confirm.clone();
+
+        glib::spawn_future_local(clone!(@weak pdf_page_start_row, @weak pdf_page_end_row, @weak input_file, @weak canvas => async move {
             let page_range =
                 (pdf_page_start_row.value() as u32 - 1)..pdf_page_end_row.value() as u32;
-            let (bytes, _) = input_file.load_bytes_future().await?;
-            canvas
-                .load_in_pdf_bytes(bytes.to_vec(), target_pos, Some(page_range))
-                .await?;
-            Ok(true)
-        }
-        _ => {
-            // Cancel
-            Ok(false)
+
+            let (bytes, _) = match input_file.load_bytes_future().await {
+                Ok(res) => {res}
+                Err(err) => {
+                    if let Err(e) = inner_tx_confirm.unbounded_send(Err(err.into())) {
+                        tracing::error!("Failed to load file, but failed to send signal through channel. Err: {e:?}");
+                    }
+                    return;
+                }
+            };
+            if let Err(e) = canvas.load_in_pdf_bytes(bytes.to_vec(), target_pos, Some(page_range)).await {
+                if let Err(e) = inner_tx_confirm.unbounded_send(Err(e)) {
+                    tracing::error!("Failed to load PDF, but failed to send signal through channel. Err: {e:?}");
+                }
+                return;
+            }
+
+            if let Err(e) = inner_tx_confirm.unbounded_send(Ok(true)) {
+                tracing::error!("PDF file imported, but failed to send signal through channel. Err: {e:?}");
+            }
+        }));
+    }));
+
+    // Overwrite builtin close shortcut
+    let controller = ShortcutController::new();
+    controller.add_shortcut(Shortcut::new(
+        Some(ShortcutTrigger:: parse_string("Escape").unwrap()),
+        Some(CallbackAction::new(clone!(@weak import_pdf_button_cancel => @default-return glib::Propagation::Stop, move |_, _| {
+            import_pdf_button_cancel.emit_clicked();
+
+            glib::Propagation::Stop
+        }))),
+    ));
+    dialog.add_controller(controller);
+
+    // Wait for a response from the dialog
+    loop {
+        if let Some(res) = rx.next().await {
```

Comment:
Is the loop really needed here? When the channel resolves to `None` the sender side has been closed, and we should return as well.

Reply:
Is there any type of error that I should return in this case?

Reply:
I guess an anyhow Error with a descriptive message should suffice. It only ends up being printed as a log message anyway